src/transformStateWithClones.js

```diff
@@ -7,7 +7,37 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const statesArray = [];
+
+  let currentState = { ...state };
+
+  actions.forEach((action) => {
+    switch (action.type) {
+      case 'clear':
+        currentState = {};
+        break;
+
+      case 'addProperties':
+        currentState = { ...currentState, ...action.extraData };
+        break;
+
+      case 'removeProperties':
+        const copy = { ...currentState };
+        action.keys.forEach((key) => {
+          delete copy[key];
+        });
+
+        currentState = { ...copy };
+        break;
+
+      default:
+        break;
```

Comment:
In the default case of the switch statement, consider throwing an error or handling unexpected action types to make the function more robust against incorrect inputs.